tot/models.py

```diff
@@ -225,3 +231,302 @@ def maybe_drop_added_dates(self, predicted, df):
+            df : pd.DataFrame
+                dataframe containing column ``ds``, ``y``, and optionally ``ID`` with all data
+            freq : str
+                frequency of the input data
+        """
+        df, received_ID_col, received_single_time_series, received_dict, _ = df_utils.prep_or_copy_df(df)
+        # Receives df with single ID column. Only single time series accepted.
+        assert len(df["ID"].unique()) == 1  # TODO: add multi-ID, multi-target
+
+        self.freq = freq
+
+    def predict(self, df: pd.DataFrame):
+        """Runs the model to make predictions.
+        Expects all data to be present in dataframe.
+
+        Parameters
+        ----------
+            df : pd.DataFrame
+                dataframe containing column ``ds``, ``y``, and optionally ``ID`` with data
+        Returns
+        -------
+            pd.DataFrame
+                columns ``ds``, ``y``, optionally [``ID``], and [``yhat<i>``] where yhat<i> refers to the
+                i-step-ahead prediction for this row's datetime, e.g. yhat3 is the prediction for this datetime,
+                predicted 3 steps ago, "3 steps old".
+
+                Note
+                ----
+                 *  raw data is not supported
+        """
+        df, received_ID_col, received_single_time_series, received_dict, _ = df_utils.prep_or_copy_df(df)
+        # Receives df with single ID column. Only single time series accepted.
+        assert len(df["ID"].unique()) == 1  # TODO: add multi-ID, multi-target
+
+        forecast = pd.DataFrame
+        # check also no id column
+        for df_name, df_i in df.groupby("ID"):
+            dates, predicted = self._predict_raw(df_i)
+            forecast = self._reshape_raw_predictions_to_forecst_df(df_i, predicted)
+        fcst_df = df_utils.return_df_in_original_format(
+            forecast, received_ID_col, received_single_time_series, received_dict
+        )
+        return fcst_df
+
+    def maybe_add_first_inputs_to_df(self, df_train, df_test):
+        """Adds last K values from df_train to start of df_test.
+
+        Parameters
+        ----------
+            df_train: pd.DataFrame
+                dataframe containing train data
+            df_test: pd.DataFrame
+                dataframe containing test data of previous split
+
+        Returns
+        -------
+            pd.DataFrame
+                dataframe containing test data enlarged with K values.
+        """
+        if self.K > 0:
+            df_train, _, _, _, _ = df_utils.prep_or_copy_df(df_train)
+            (
+                df_test,
+                received_ID_col_test,
+                received_single_time_series_test,
+                received_dict_test,
+                _,
+            ) = df_utils.prep_or_copy_df(df_test)
+            df_test_new = pd.DataFrame()
+            for df_name, df_test_i in df_test.groupby("ID"):
+                df_train_i = df_train[df_train["ID"] == df_name].copy(deep=True)
```

Comment:
maybe only copying the tail may be sufficient?

Reply:
@ourownstory do you mean the tail of df_train or df_test? 
For as single-ID df copying the tail of df_train would not be sufficient anymore. However, in case we have multiple IDs, copying the tail of df_train would not be sufficient anymore.